### ISSUE TITLE
Weird lockups in PC2

### ISSUE THREAD

**Issue by karalabe** [filecoin-project/rust-fil-proofs#1333]
I've seen this countless times now and I'm not sure where to put it. Every now and again, a PC2 run just locks up for a seemingly random amount of time (minutes to hours) and then just picks up as if nothing happened.

E.g.:

```
Oct 29 13:19:43 filecoin lotus-worker[10769]: {"level":"info","ts":"2020-10-29T13:19:43.021+0200","logger":"storage_proofs_porep::stacked::vanilla::proof","caller":"/root/.cargo/registry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:429","msg":"generating tree c using the GPU"}

[... 22 minute hang... ]

Oct 29 13:41:28 filecoin lotus-worker[10769]: {"level":"info","ts":"2020-10-29T13:41:28.615+0200","logger":"storage_proofs_porep::stacked::vanilla::proof","caller":"/root/.cargo/registry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 1/8 of length 153391689"}
Oct 29 13:44:03 filecoin lotus-worker[10769]: {"level":"info","ts":"2020-10-29T13:44:03.268+0200","logger":"storage_proofs_porep::stacked::vanilla::proof","caller":"/root/.cargo/registry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 2/8 of length 153391689"}
Oct 29 13:46:38 filecoin lotus-worker[10769]: {"level":"info","ts":"2020-10-29T13:46:38.725+0200","logger":"storage_proofs_porep::stacked::vanilla::proof","caller":"/root/.cargo/registry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 3/8 of length 153391689"}
Oct 29 13:49:29 filecoin lotus-worker[10769]: {"level":"info","ts":"2020-10-29T13:49:29.506+0200","logger":"storage_proofs_porep::stacked::vanilla::proof","caller":"/root/.cargo/registry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 4/8 of length 153391689"}
```

You can see that the layers are super faster afterwards. The GPU was idle in the mean time.

Another example:

```
Oct 29 12:55:31 filecoin lotus-worker[10769]: {"level":"info","ts":"2020-10-29T12:55:31.095+0200","logger":"storage_proofs_porep::stacked::vanilla::proof","caller":"/root/.cargo/registry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 7/8 of length 153391689"}
Oct 29 13:14:29 filecoin lotus-worker[10769]: {"level":"info","ts":"2020-10-29T13:14:29.994+0200","logger":"storage_proofs_porep::stacked::vanilla::proof","caller":"/root/.cargo/registry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 8/8 of length 153391689"}
Oct 29 13:15:14 filecoin lotus-worker[10769]: {"level":"info","ts":"2020-10-29T13:15:14.089+0200","logger":"storage_proofs_porep::stacked::vanilla::proof","caller":"/root/.cargo/registry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:1061","msg":"tree_c done"}
Oct 29 13:15:14 filecoin lotus-worker[10769]: {"level":"info","ts":"2020-10-29T13:15:14.089+0200","logger":"storage_proofs_porep::stacked::vanilla::proof","caller":"/root/.cargo/registry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:1088","msg":"building tree_r_last"}
Oct 29 13:15:14 filecoin lotus-worker[10769]: {"level":"info","ts":"2020-10-29T13:15:14.090+0200","logger":"storage_proofs_porep::stacked::vanilla::proof","caller":"/root/.cargo/registry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:727","msg":"generating tree r last using the GPU"}
Oct 29 13:16:59 filecoin lotus-worker[10769]: {"level":"info","ts":"2020-10-29T13:16:59.342+0200","logger":"storage_proofs_porep::stacked::vanilla::proof","caller":"/root/.cargo/registry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:826","msg":"building base tree_r_last with GPU 1/8"}

[... stuck, no output for 40 minutes now... ]

EDIT: the 2/8 appeared at 14:07:04, so about 50 minutes after the 1/8, then it picked up and finished the rest with normal stream, concurrently with the other PC2 (also at normal speed).
```

One thing that might maybe shed some light is that the first trace above started exactly where the second left off. Seems these concurrent PC2 runs influence each other somehow? There's enough video memory for both, so that can't be. And even if the former suspends the latter, it does not explain why the former above is stuck for 20 mins with no output, no gpu load, no disk load nothing.

**Comment by karalabe** [comment]
For example this morning I had a PC2 stuck for 7 hours, another for 3 and another for 1.5 when I just restarted the process. Then half of them finished within 25 minutes.

**Comment by porcuquine** [comment]
Oh, you are trying to run multiple PC2 at once on the same GPU? I am not surprised that produces weird behavior. There is basically no GPU scheduling now, and all the workloads are designed around the assumption of one batch at a time. This is likely confusing the GPU memory management in unpredictable ways. Getting something better figured out is pretty high priority, and maybe I can come up with something intermediate as a stopgap. But I would say that if you can avoid doing that, it's likely to make your life… better.

**Comment by porcuquine** [comment]
Even just interleaving batches from multiple jobs should be fine, not that this is necessarily actionable for you. I'm just thinking out loud.

**Comment by karalabe** [comment]
Well, to be fair, I **can** run 5 PC2 concurrently on a single card  at maybe a x2 runtime penalty (so 5/2 is still 2.5 more throughput).

```
ry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 2/8 of length 153391689"}
ry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 2/8 of length 153391689"}
ry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 2/8 of length 153391689"}
ry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 1/8 of length 153391689"}
ry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 1/8 of length 153391689"}
ry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 3/8 of length 153391689"}
ry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 3/8 of length 153391689"}
ry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 3/8 of length 153391689"}
ry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 4/8 of length 153391689"}
ry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 4/8 of length 153391689"}
ry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 4/8 of length 153391689"}
ry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 2/8 of length 153391689"}
ry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 2/8 of length 153391689"}
ry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 5/8 of length 153391689"}
ry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 5/8 of length 153391689"}
ry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 5/8 of length 153391689"}
ry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 6/8 of length 153391689"}
ry/src/github.com-1ecc6299db9ec823/storage-proofs-porep-5.2.0/src/stacked/vanilla/proof.rs:560","msg":"persisting base tree_c 6/8 of length 153391689"}
```

(the slower ones above are mostly due to being served from HDD instead of SSD). I really think you should make sure PC2s don't interfere with each other as they allow a lot of optimizations. As long as there's enough video RAM for all.

**Comment by porcuquine** [comment]
On further consideration, I think the serialization I suggested is already enforced, and this sounds consistent with what you are reporting. I suspect something else is blocking the process in your case.

When you are running multiple (like 5) PC2 'concurrently on a single card', do you observe an increase in GPU RAM usage? I think you shouldn't, but if you are then that would be useful to know (in as much detail as you can provide). For example, memory usage is X with 1 process, Y with 2, etc.

**Comment by porcuquine** [comment]
Trying to hash multiple batches at once on a single GPU is definitely a bad idea. It won't help, either, because the job is compute bound. This was meant to have been enforced but slipped through the cracks and should be fixed here: #1335 

I would expect much less weird behavior, including memory leaks (transient bloat, or otherwise).

**Comment by cryptonemo** [comment]
I believe this issue has been resolved.  Please open a new issue if not.
